src/runtime/backing.rs

```diff
@@ -0,0 +1,322 @@
+    //                     Some(ImportValue::Table(t)) => t.to_vec(),
+    //                     None => {
+    //                         if options.mock_missing_tables {
+    //                             debug!(
+    //                                 "The Imported table {}.{} is not provided, therefore will be mocked.",
+    //                                 module_name, field_name
+    //                             );
+    //                             let len = table.entity.minimum as usize;
+    //                             let mut v = Vec::with_capacity(len);
+    //                             v.resize(len, 0);
+    //                             v
+    //                         } else {
+    //                             panic!(
+    //                                 "Imported table value was not provided ({}.{})",
+    //                                 module_name, field_name
+    //                             )
+    //                         }
+    //                     }
+    //                     _ => panic!(
+    //                         "Expected global table, but received {:?} ({}.{})",
+    //                         imported, module_name, field_name
+    //                     ),
+    //                 }
+    //             }
+    //             None => {
+    //                 let len = table.entity.minimum as usize;
+    //                 let mut v = Vec::with_capacity(len);
+    //                 v.resize(len, 0);
+    //                 v
+    //             }
+    //         };
+    //         tables.push(table);
+    //     }
+
+    //     // instantiate tables
+    //     for table_element in &module.info.table_elements {
+    //         let base = match table_element.base {
+    //             Some(global_index) => globals_data[global_index.index()] as usize,
+    //             None => 0,
+    //         };
+
+    //         let table = &mut tables[table_element.table_index.index()];
+    //         for (i, func_index) in table_element.elements.iter().enumerate() {
+    //             // since the table just contains functions in the MVP
+    //             // we get the address of the specified function indexes
+    //             // to populate the table.
+
+    //             // let func_index = *elem_index - module.info.imported_funcs.len() as u32;
+    //             // let func_addr = functions[func_index.index()].as_ptr();
+    //             let func_addr = get_function_addr(&func_index, &import_functions, &functions);
+    //             table[base + table_element.offset + i] = func_addr as _;
+    //         }
+    //     }
+    // }
+}
+
+#[derive(Debug)]
+pub struct ImportBacking {
+    pub functions: Box<[vm::ImportedFunc]>,
+    pub memories: Box<[vm::ImportedMemory]>,
+    pub tables: Box<[vm::ImportedTable]>,
+    pub globals: Box<[vm::ImportedGlobal]>,
+}
+
+impl ImportBacking {
+    pub fn new(module: &Module, imports: &Imports) -> Result<Self, String> {
```

Comment:
What you think if instead of providing `Imports` as a specific implementation, we just provide a function that given two Strings returns an optional `Import`.

Something like:
```rust
fn get_import(module: &str, field: &str) -> Option<Import> {
   // ...
}
```

Reply:
This way we abstract the real implementation of the imports, so we can use a nested Hashmap (this PR impl), a Hashmap with a pair of keys (current impl), or really any other thing.

Reply:
Like supplying it a closure that takes to `&str` args and returns an `Option<Import>`?

```rust
pub fn new(module: &Module, imports: impl Fn(&str, &str) -> Option<Import>) -> Result<Self, String? { ... }
```
Like that?

Reply:
Yup, that sounds good to me (using `import_func` instead of `imports`).
What you think?

Reply:
Yeah, that would be more tunable for specific use cases I suppose. Only issue is that monomorphization may end up blowing up compile-time and binary-size if we make it generic on a closure.